pkg/process/process_test.go

```diff
@@ -147,107 +152,229 @@ func (s *TestSuite) TestProcessDeletion(c *C) {
+// additional sents will be blocked.
+// https://github.com/longhorn/longhorn/issues/2697
+func (s *TestSuite) TestProcessReplace(c *C) {
+	count := 100
+	wg := &sync.WaitGroup{}
+	pw := &ProcessWatcher{}
+	for i := 0; i < count; i++ {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			go s.pm.ProcessWatch(nil, pw)
+			name := "test_process_replace-" + strconv.Itoa(i)
+			assertProcessCreation(c, s.pm, name, TestBinary)
+			assertProcessReplace(c, s.pm, name, TestBinaryReplace)
 
-			wgDeletion.Wait()
-			deleted := false
-			for j := 0; j < RetryCount; j++ {
-				listResp, err := s.pm.ProcessList(nil, &rpc.ProcessListRequest{})
-				c.Assert(err, IsNil)
-				_, exists := listResp.Processes[name]
-				if !exists {
-					deleted = true
-					break
-				}
-				time.Sleep(RetryInterval)
-			}
+			assertProcessDeletion(c, s.pm, name)
+			deleted, err := waitForProcessListState(s.pm, func(processes map[string]*rpc.ProcessResponse) bool {
+				_, exists := processes[name]
+				return !exists
+			})
+			c.Assert(err, IsNil)
 			c.Assert(deleted, Equals, true)
+		}(i)
+	}
+	wg.Wait()
+}
 
-			// after previous command delete the process of the same name, creating the process again
-			// and make sure it does run and exist for deletion later
-			createReq = &rpc.ProcessCreateRequest{
-				Spec: &rpc.ProcessSpec{
-					Name:      name,
-					Binary:    binary,
-					Args:      []string{},
-					PortCount: 1,
-					PortArgs:  nil,
-				},
+// there was a race in the process UpdateChannel assignment
+// this lead to a deadlock when the binary couldn't be located.
+// https://github.com/longhorn/longhorn/issues/2697
+func (s *TestSuite) TestProcessReplaceMissingBinary(c *C) {
+	count := 100
+	wg := &sync.WaitGroup{}
+	pw := &ProcessWatcher{}
+	for i := 0; i < count; i++ {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			go s.pm.ProcessWatch(nil, pw)
+			name := "test_process_missing_binary_replace-" + strconv.Itoa(i)
+			assertProcessCreation(c, s.pm, name, TestBinary)
+
+			// replacement for a missing binary should error
+			_, err := s.pm.ProcessReplace(nil, &rpc.ProcessReplaceRequest{
+				Spec:            createProcessSpec(name, TestBinaryMissing),
+				TerminateSignal: "SIGHUP",
+			})
+			c.Assert(err, NotNil)
+		}(i)
+	}
+	wg.Wait()
+}
+
+// there was a nil pointer case, while updating a process that is being
+// deleted, since when initially checked the process was still in the map
+// but by the time new process has started the old process had been removed
+func (s *TestSuite) TestProcessReplaceDuringDeletion(c *C) {
+	count := 100
+	wg := &sync.WaitGroup{}
+	pw := &ProcessWatcher{}
+	for i := 0; i < count; i++ {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			go s.pm.ProcessWatch(nil, pw)
+			name := "test_process_deletion_replace-" + strconv.Itoa(i)
+			assertProcessCreation(c, s.pm, name, TestBinary)
+			assertProcessDeletion(c, s.pm, name)
```

Comment:
When this deletion is verified, the old process is already not in the map when the initial check of the process replacement happens. Then does this test still make sense?

Reply:
Is a required test, since deletion of the process from the map happens async.
The process deletion is triggered, but not removed from the map till the process has actually successfully stopped.

See: https://github.com/joshimoo/longhorn-instance-manager/blob/9781dad48aa70fd86e1a948dc34c4bc7492b53c2/pkg/process/process_manager.go#L233

This is one problem with the way the processes are handled in the IM currently.
Another problem is the port allocations are not bound to a specific process UUID, so it's possible that one process releases/claims anothers ports.

But for v1.1.2 I did not want to touch the process structures, instead I used these tests to verifiy the expected functionality based on the current design, which I have already identified as flawed.